source/common/upstream/original_dst_cluster.cc

```diff
@@ -166,48 +167,76 @@ OriginalDstCluster::OriginalDstCluster(
 }
 
 void OriginalDstCluster::addHost(HostSharedPtr& host) {
-  HostMapSharedPtr new_host_map = std::make_shared<HostMap>(*getCurrentHostMap());
-  auto pair = new_host_map->emplace(host->address()->asString(), host);
-  bool added = pair.second;
-  if (added) {
-    ENVOY_LOG(debug, "addHost() adding {}", host->address()->asString());
-    setHostMap(new_host_map);
-    // Given the current config, only EDS clusters support multiple priorities.
-    ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
-    const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
-    HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
-    all_hosts->emplace_back(host);
-    priority_set_.updateHosts(0,
-                              HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
-                              {}, {std::move(host)}, {}, absl::nullopt);
+  std::string address = host->address()->asString();
+  HostMultiMapSharedPtr new_host_map = std::make_shared<HostMultiMap>(*getCurrentHostMap());
+  auto it = new_host_map->find(address);
+  if (it != new_host_map->end()) {
+    it->second->hosts_.push_back(host);
+  } else {
+    new_host_map->emplace(address, std::make_shared<HostsForAddress>(host));
   }
+  ENVOY_LOG(debug, "addHost() adding {} {}.", *host, address);
+  setHostMap(new_host_map);
+
+  // Given the current config, only EDS clusters support multiple priorities.
+  ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
+  const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
+  HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
+  all_hosts->emplace_back(host);
+  priority_set_.updateHosts(0,
+                            HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
+                            {}, {std::move(host)}, {}, absl::nullopt);
 }
 
 void OriginalDstCluster::cleanup() {
   HostVectorSharedPtr keeping_hosts(new HostVector);
   HostVector to_be_removed;
-  ENVOY_LOG(trace, "Stale original dst hosts cleanup triggered.");
+  absl::flat_hash_set<absl::string_view> removed_addresses;
   auto host_map = getCurrentHostMap();
   if (!host_map->empty()) {
     ENVOY_LOG(trace, "Cleaning up stale original dst hosts.");
-    for (const auto& pair : *host_map) {
-      const std::string& addr = pair.first;
-      const HostSharedPtr& host = pair.second;
-      if (host->used()) {
-        ENVOY_LOG(trace, "Keeping active host {}.", addr);
-        keeping_hosts->emplace_back(host);
-        host->used(false); // Mark to be removed during the next round.
+    for (const auto& [addr, hosts] : *host_map) {
+      // Address is kept in the cluster if either of the two things happen:
+      // 1) a host has been recently selected for the address; 2) none of the
+      // hosts are currently in any of the connection pools.
+      // The set of hosts for a single address are treated as a unit.
```

Comment:
Sorry I still don't understand why we are doing this? Can you add more comments? Is this just for legacy behavior or is there some other reason? Basically can the used_ thing be removed once the feature flag is removed?

Reply:
Here's a sequence I'm worried about:
1) worker1 picks a host H in `chooseHost`
2) main runs cleanup that calls drain on H
3) worker1 creates a handle for H

If we continue using this `used_` bool then 2) would not happen between 1)-3), as long as time between 1)-3) is less than `cleanup_interval` (which it is).

Happy to hear any other way around it? It doesn't really hurt to preserve the legacy behavior AFAICT.

Reply:
> If we continue using this used_ bool then 2) would not happen between 1)-3), as long as time between 1)-3) is less than cleanup_interval (which it is).

How does the time interval matter here? If time matters I think it could still race? Or the boolean will prevent cleanup?

I guess in general I'm still thinking that switching to locking would be easier to reason about and it's unclear to me that there will be much/any perf delta?

Reply:
The boolean adds a minimum delay (seconds?). It'll take a really slow execution to execute these two lines with seconds delay:
```
auto h = lb.chooseHost(&context);
auto handle = h.acquireHandle();
```

Reply:
If we want to use locking, we'd have to rewrite most of this cluster. That is likely not backwards compatible in one way or another, so I'm worried that it's better to create another cluster with better characteristics instead.

Reply:
Alright fair enough. Let's stick with this.